static/js/components/widgets/HierarchicalSelectField.tsx

```diff
@@ -0,0 +1,146 @@
+import { sortedUniqBy, sortBy, times } from "lodash"
+
+import SelectField from "./SelectField"
+
+type Level = { label: string; name: string }
+type Option = { label: string; value: string }
+type OptionsMap = Record<string, Record<string, any> | string[]>
+type Props = {
+  name: string
+  label: string
+  label_singular?: string // eslint-disable-line camelcase
+  levels: Level[] // eslint-disable-line camelcase
+  value: Tuple[] | null
+  onChange: (event: any) => void
+  options_map: OptionsMap // eslint-disable-line camelcase
+}
+
+const asOption = (str: string | null) => ({
+  label: str ?? "-- empty --",
+  value: str ?? ""
+})
+
+type TupleItem = string | null
+type Tuple = TupleItem[]
+
+export const describeTuple = (tuple: Tuple): string =>
+  tuple.filter(Boolean).join(" - ")
+
+export const calcOptions = (
+  optionsMap: OptionsMap,
+  selectedTuple: Tuple,
+  levels: Level[]
+): Option[][] => {
+  let previousSelected = true
+  let currentOptionsMap: any = optionsMap
+
+  return levels.map((_, levelIdx) => {
+    const selectedItem = selectedTuple[levelIdx]
+    if (!previousSelected) {
+      return [null].map(asOption)
+    }
+
+    if (!selectedItem) {
+      previousSelected = false
+    }
+
+    let optionValues: Array<string | null>
+    if (Array.isArray(currentOptionsMap)) {
+      optionValues = [null, ...sortBy(currentOptionsMap)]
+      currentOptionsMap = null
+    } else {
+      const keys = Object.keys(currentOptionsMap)
+      // @ts-ignore
+      optionValues = [null, ...sortBy(keys)]
+      currentOptionsMap = selectedItem ? currentOptionsMap[selectedItem] : null
+    }
+    return optionValues.map(asOption)
+  })
+}
+
+export default function HierarchicalSelectField(props: Props): JSX.Element {
+  const { options_map: optionsMap, levels, value, onChange, name } = props
+
+  const labelSingular = props.label_singular ?? props.label
+  const valueList = value ?? []
+
+  const defaultValue = times(levels.length, () => null)
+  const [selectedTuple, setSelectedTuple] = useState<Tuple>(defaultValue)
+
+  const setSelectedValueFor = (levelIdx: number, selected: string) => {
+    setSelectedTuple(oldTuple =>
+      oldTuple.map((tuple, idx) =>
+        idx < levelIdx ? tuple : idx === levelIdx ? selected : null
+      )
+    )
+  }
+
+  const handleAdd = (event: any) => {
+    event.preventDefault()
+
+    if (selectedTuple[0] === null) {
+      // nothing is selected, ignore
+      return
+    }
+
+    const combinedList = [selectedTuple, ...valueList]
+    onChange({
+      target: {
+        value: sortedUniqBy(sortBy(combinedList, describeTuple), describeTuple),
+        name
+      }
+    })
+    setSelectedTuple(defaultValue)
+  }
+
+  const handleDelete = (idx: number) => (event: any) => {
+    event.preventDefault()
+
+    onChange({
+      target: {
+        value: valueList.filter((_, _idx) => _idx !== idx),
+        name
+      }
+    })
+  }
+
+  const options = calcOptions(optionsMap, selectedTuple, levels)
```

Comment:
mmm just thought about this a little more and realized it's not so simple, just wondering about how that `calcOptions` function could maybe be made simpler somehow...

Reply:
possibly could be helpful to just add a few little comments explaining how it works?

Reply:
Sorry, this is a little convoluted. `calcOptions` essentially returns an `[Option[], Option[], Option[]]`, one per topic, subtopic, and specialty given the field data and the currently selected items in `selectedTuple`. It iterates through the levels:
 - `null` represents the empty item at the beginning of the list and means no selection. Every set of options must have a `null` at the beginning of the list
 - if there is nothing selected at that level, none of the more specific selects are selected either, since for example you can't have a subtopic selected and not have a topic selected
 - The structure of the `options_map` is nested `dict`s and a `list` at the innermost layer. The keys of the `dict` are used, otherwise it's the elements of the list.

Reply:
I updated the function to simplify a little and add comments